// @waves/waves-crypto/dist/index.js
"use strict";
// Copyright (c) 2018 Yuriy Naydenov
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
const CryptoJS = require("crypto-js");
const blake = require("./libs/blake2b");
const sha3_1 = require("./libs/sha3");
const base58_1 = require("./libs/base58");
const axlsign_1 = require("./libs/axlsign");
exports.libs = {
    CryptoJS,
    blake,
    keccak256: sha3_1.keccak256,
    base58: base58_1.default,
    axlsign: axlsign_1.default
};
exports.concat = (...arrays) => arrays.reduce((a, b) => Uint8Array.from([...a, ...b]), new Uint8Array(0));
function buildAddress(publicKeyBytes, chainId = 'W') {
    const prefix = [1, chainId.charCodeAt(0)];
    const publicKeyHashPart = hashChain(publicKeyBytes).slice(0, 20);
    const rawAddress = exports.concat(prefix, publicKeyHashPart);
    const addressHash = Uint8Array.from(hashChain(rawAddress).slice(0, 4));
    return base58_1.default.encode(exports.concat(rawAddress, addressHash));
}
exports.buildAddress = buildAddress;
function buildSeedHash(seedBytes, nonce) {
    const nonceArray = [0, 0, 0, 0];
    if (nonce && nonce > 0) {
        let remainder = nonce;
        for (let i = 3; i >= 0; i--) {
            nonceArray[3 - i] = Math.floor(remainder / Math.pow(2, (i * 8)));
            remainder = remainder % Math.pow(2, (i * 8));
        }
    }
    const seedBytesWithNonce = exports.concat(nonceArray, seedBytes);
    const seedHash = hashChain(seedBytesWithNonce);
    return sha256(seedHash);
}
exports.buildSeedHash = buildSeedHash;
function byteArrayToWordArrayEx(arr) {
    const len = arr.length;
    const words = [];
    for (let i = 0; i < len; i++) {
        words[i >>> 2] |= (arr[i] & 0xff) << (24 - (i % 4) * 8);
    }
    return CryptoJS.lib.WordArray.create(words, len);
}
function wordArrayToByteArrayEx(wordArray) {
    let words = wordArray.words;
    let sigBytes = wordArray.sigBytes;
    let u8 = new Uint8Array(sigBytes);
    for (let i = 0; i < sigBytes; i++) {
        let byte = (words[i >>> 2] >>> (24 - (i % 4) * 8)) & 0xff;
        u8[i] = byte;
    }
    return u8;
}
exports.stringToUint8Array = (str) => Uint8Array.from([...unescape(encodeURIComponent(str))].map(c => c.charCodeAt(0)));
exports.publicKeyToString = (pk) => typeof pk === 'string' ? pk : (pk instanceof Uint8Array ? exports.base58encode(pk) : pk.public);
exports.ADDRESS_LENGTH = 26;
exports.PUBLIC_KEY_LENGTH = 32;
exports.PRIVATE_KEY_LENGTH = 32;
exports.SIGNATURE_LENGTH = 64;
function blake2b(input) {
    return blake.blake2b(input, null, 32);
}
exports.blake2b = blake2b;
function keccak(input) {
    return sha3_1.keccak256.array(input);
}
exports.keccak = keccak;
function sha256(input) {
    const wordArray = byteArrayToWordArrayEx(input);
    const resultWordArray = CryptoJS.SHA256(wordArray);
    return wordArrayToByteArrayEx(resultWordArray);
}
exports.sha256 = sha256;
function hashChain(input) {
    return Uint8Array.from(keccak(blake2b(input)));
}
exports.base58encode = (input) => base58_1.default.encode(input);
exports.base58decode = (input) => base58_1.default.decode(input);
exports.keyPair = (seed) => {
    const seedBytes = exports.stringToUint8Array(seed);
    const seedHash = buildSeedHash(seedBytes);
    const keys = axlsign_1.default.generateKeyPair(seedHash);
    return {
        private: base58_1.default.encode(keys.private),
        public: base58_1.default.encode(keys.public)
    };
};
exports.publicKey = (seed) => exports.keyPair(seed).public;
exports.privateKey = (seed) => exports.keyPair(seed).private;
exports.address = (keyOrSeed, chainId = 'W') => typeof keyOrSeed === 'string' ?
    exports.address(exports.keyPair(keyOrSeed), chainId) :
    buildAddress(base58_1.default.decode(keyOrSeed.public), chainId);
exports.signBytes = (bytes, seed) => exports.signWithPrivateKey(bytes, exports.privateKey(seed));
exports.verifySignature = (publicKey, bytes, signature) => {
    const signatureBytes = base58_1.default.decode(signature);
    return (signatureBytes.length == exports.SIGNATURE_LENGTH &&
        axlsign_1.default.verify(base58_1.default.decode(publicKey), bytes, signatureBytes));
};
function arraysEqual(a, b) {
    if (a === b)
        return true;
    if (a == null || b == null)
        return false;
    if (a.length != b.length)
        return false;
    for (var i = 0; i < a.length; ++i)
        if (a[i] !== b[i])
            return false;
    return true;
}
exports.arraysEqual = arraysEqual;
exports.hashBytes = (bytes) => base58_1.default.encode(blake2b(bytes));
exports.signWithPrivateKey = (dataBytes, privateKey) => {
    const privateKeyBytes = base58_1.default.decode(privateKey);
    const signature = axlsign_1.default.sign(privateKeyBytes, dataBytes, randomUint8Array(64));
    return base58_1.default.encode(signature);
};
function nodeRandom(count, options) {
    const crypto = require('crypto');
    const buf = crypto.randomBytes(count);
    switch (options.type) {
        case 'Array':
            return [].slice.call(buf);
        case 'Buffer':
            return buf;
        case 'Uint8Array':
            return Uint8Array.from(buf);
        default:
            throw new Error(options.type + ' is unsupported.');
    }
}
function browserRandom(count, options) {
    const nativeArr = new Uint8Array(count);
    const crypto = global.crypto || global.msCrypto;
    crypto.getRandomValues(nativeArr);
    switch (options.type) {
        case 'Array':
            return [].slice.call(nativeArr);
        case 'Buffer':
            try {
                const b = new Buffer(1);
            }
            catch (e) {
                throw new Error('Buffer not supported in this environment. Use Node.js or Browserify for browser support.');
            }
            return new Buffer(nativeArr);
        case 'Uint8Array':
            return nativeArr;
        default:
            throw new Error(options.type + ' is unsupported.');
    }
}
const isBrowser = typeof window !== 'undefined' && ({}).toString.call(window) === '[object Window]';
const isNode = typeof global !== 'undefined' && ({}).toString.call(global) === '[object global]';
const isJest = process.env.JEST_WORKER_ID !== undefined;
function secureRandom(count, options) {
    options = options || { type: 'Array' };
    if (isBrowser) {
        return browserRandom(count, options);
    }
    else if (isNode) {
        return nodeRandom(count, options);
    }
    else if (isJest) {
        return nodeRandom(count, options);
    }
    else {
        throw new Error('Your environment is not defined');
    }
}
function randomUint8Array(length) {
    return secureRandom(length, { type: 'Uint8Array' });
}
exports.randomUint8Array = randomUint8Array;
let charToNibble = {};
let nibbleToChar = [];
let i;
for (i = 0; i <= 9; ++i) {
    let character = i.toString();
    charToNibble[character] = i;
    nibbleToChar.push(character);
}
for (i = 10; i <= 15; ++i) {
    let lowerChar = String.fromCharCode('a'.charCodeAt(0) + i - 10);
    let upperChar = String.fromCharCode('A'.charCodeAt(0) + i - 10);
    charToNibble[lowerChar] = i;
    charToNibble[upperChar] = i;
    nibbleToChar.push(lowerChar);
}
function byteArrayToHexString(bytes) {
    let str = '';
    for (let i = 0; i < bytes.length; ++i) {
        if (bytes[i] < 0) {
            bytes[i] += 256;
        }
        str += nibbleToChar[bytes[i] >> 4] + nibbleToChar[bytes[i] & 0x0F];
    }
    return str;
}
exports.byteArrayToHexString = byteArrayToHexString;
function hexStringToByteArray(str) {
    let bytes = [];
    let i = 0;
    if (0 !== str.length % 2) {
        bytes.push(charToNibble[str.charAt(0)]);
        ++i;
    }
    for (; i < str.length - 1; i += 2)
        bytes.push((charToNibble[str.charAt(i)] << 4) + charToNibble[str.charAt(i + 1)]);
    return bytes;
}
exports.hexStringToByteArray = hexStringToByteArray;
function getSharedKey(privateKeyFrom, publicKeyTo) {
    const prvk = exports.base58decode(privateKeyFrom);
    const pubk = exports.base58decode(publicKeyTo);
    return axlsign_1.default.sharedKey(prvk, pubk);
}
exports.getSharedKey = getSharedKey;
function getKEK(sharedKey, prefix = 'waves') {
    let bytesSharedKey = null;
    try {
        bytesSharedKey = exports.base58decode(sharedKey);
        if (!bytesSharedKey || bytesSharedKey.length < 32) {
            throw new Error('Invalid sharedKey length');
        }
    }
    catch (e) {
        throw e;
    }
    const P = exports.stringToUint8Array(prefix);
    const KEK_Bytes = new Uint8Array(bytesSharedKey.length + P.length);
    KEK_Bytes.set(bytesSharedKey);
    KEK_Bytes.set(P, bytesSharedKey.length);
    const KEK = CryptoJS.SHA256(byteArrayToWordArrayEx(KEK_Bytes));
    return {
        KEK,
        P,
    };
}
exports.getKEK = getKEK;
function encryptMessage(sharedKey, message, prefix = 'waves') {
    const { KEK, P } = getKEK(sharedKey, prefix);
    const M = CryptoJS.enc.Utf8.parse(message);
    const CEK_Bytes = randomUint8Array(32);
    const CEK = byteArrayToWordArrayEx(CEK_Bytes);
    const IV_Bytes = randomUint8Array(16);
    const CEK_FOR_HMAC = CEK_Bytes.map((byte, index) => byte | P[index % P.length]);
    const Cc = CryptoJS.AES.encrypt(M, CEK, {
        iv: byteArrayToWordArrayEx(IV_Bytes),
        mode: CryptoJS.mode.CTR,
    });
    const C = CryptoJS.enc.Base64.parse(Cc.toString());
    const Mhmac = CryptoJS.HmacSHA256(M, CEK);
    const Ccek = CryptoJS.enc.Base64.parse(CryptoJS.AES.encrypt(CEK, KEK, { mode: CryptoJS.mode.ECB }).toString());
    const CEKhmac = CryptoJS.HmacSHA256(byteArrayToWordArrayEx(CEK_FOR_HMAC), KEK);
    const CcekBytes = wordArrayToByteArrayEx(Ccek);
    const CEKhmacBytes = wordArrayToByteArrayEx(CEKhmac);
    const CBytes = wordArrayToByteArrayEx(C);
    const MhmacBytes = wordArrayToByteArrayEx(Mhmac);
    const packageBytes = new Uint8Array(CcekBytes.length + CEKhmacBytes.length + CBytes.length + MhmacBytes.length + IV_Bytes.length);
    packageBytes.set(CcekBytes);
    packageBytes.set(CEKhmacBytes, CcekBytes.length);
    packageBytes.set(CBytes, CcekBytes.length + CEKhmacBytes.length);
    packageBytes.set(MhmacBytes, CcekBytes.length + CEKhmacBytes.length + CBytes.length);
    packageBytes.set(IV_Bytes, CcekBytes.length + CEKhmacBytes.length + CBytes.length + MhmacBytes.length);
    return CryptoJS.enc.Base64.stringify(byteArrayToWordArrayEx(packageBytes));
}
exports.encryptMessage = encryptMessage;
function decryptMessage(sharedKey, encryptedMessage, prefix = 'waves') {
    const { KEK, P } = getKEK(sharedKey, prefix);
    const packageBytes = wordArrayToByteArrayEx(CryptoJS.enc.Base64.parse(encryptedMessage));
    const IV_Bytes = packageBytes.slice(-16);
    const MhmacBytes = packageBytes.slice(-(16 + 32), -16);
    const CcekBytes = packageBytes.slice(0, 48);
    const CEKhmacBytes = packageBytes.slice(48, 48 + 32);
    const CBytes = packageBytes.slice(48 + 32, -(16 + 32));
    const CEK = byteArrayToWordArrayEx(wordArrayToByteArrayEx(CryptoJS.AES.decrypt(CryptoJS.enc.Base64.stringify(byteArrayToWordArrayEx(CcekBytes)), KEK, { mode: CryptoJS.mode.ECB })));
    const CEK_FOR_HMAC = wordArrayToByteArrayEx(CEK).map((byte, index) => byte | P[index % P.length]);
    const CEKhmac = wordArrayToByteArrayEx(CryptoJS.HmacSHA256(byteArrayToWordArrayEx(CEK_FOR_HMAC), KEK));
    const isValidKey = CEKhmac.every((v, i) => v === CEKhmacBytes[i]);
    if (!isValidKey) {
        throw new Error('Invalid message');
    }
    const M = CryptoJS.AES.decrypt(CryptoJS.enc.Base64.stringify(byteArrayToWordArrayEx(CBytes)), CEK, {
        iv: byteArrayToWordArrayEx(IV_Bytes),
        mode: CryptoJS.mode.CTR,
    });
    const Mhmac = wordArrayToByteArrayEx(CryptoJS.HmacSHA256(byteArrayToWordArrayEx(wordArrayToByteArrayEx(M)), CEK));
    const isValidMessage = Mhmac.every((v, i) => v === MhmacBytes[i]);
    if (!isValidMessage) {
        throw new Error('Invalid message');
    }
    return M.toString(CryptoJS.enc.Utf8);
}
exports.decryptMessage = decryptMessage;
